### PR TITLE
Manage the contract execution thread generically from the runtime

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

Another step towards running contracts on a single thread (#1193)

## Proposal

*  Make contract trait sync and create the contract execution thread in the runtime

## Test Plan

CI

### CODE REVIEW

File: linera-execution/src/execution.rs [hunk 1]
```diff
@@ -215,9 +215,10 @@ where
             UserAction::Message(context, message) => {
                 contract.execute_message(context, runtime_sender, message)
             }
-        };
+        });
         let (runtime_result, call_result) =
             futures::future::join(runtime_actor.run(), call_result_future).await;
+        let call_result = call_result?;
```
Comment (jvff) [inline linera-execution/src/execution.rs:221]: Nit: I think the join isn't needed anymore. Now you can do:

```rust
let runtime_result = runtime_actor.run().await;
call_result_future.await?;
runtime_result?;
```

File: linera-execution/src/execution.rs [hunk 2]
```diff
@@ -388,11 +389,13 @@ where
...
                 // Run the query.
-                let response_future = service.handle_query(context, runtime_sender, bytes);
+                let response_future = tokio::task::spawn_blocking(move || {
+                    service.handle_query(context, runtime_sender, bytes)
+                });
                 let (runtime_result, response) =
                     futures::future::join(runtime_actor.run(), response_future).await;
                 runtime_result?;
-                let response = response?;
+                let response = response??;
```
Comment (jvff) [inline linera-execution/src/execution.rs:398]: Nit: Same as above.

File: linera-execution/src/execution.rs [hunk 3]
```diff
@@ -216,10 +216,8 @@ where
                 contract.execute_message(context, runtime_sender, message)
             }
         });
-        let (runtime_result, call_result) =
-            futures::future::join(runtime_actor.run(), call_result_future).await;
-        let call_result = call_result?;
-        runtime_result?;
+        runtime_actor.run().await?;
+        let call_result = call_result_future.await?;
```
Comment (ma2bd) [inline linera-execution/src/execution.rs:220]: I'm assuming that any error in `runtime_actor` will cause a less precise error in `contract.execute_*`.

File: linera-execution/src/execution.rs (hunk 1 above)
Comment (ma2bd) [reply linera-execution/src/execution.rs]: good catch!

**Review by jvff — APPROVED** [review]

File: linera-execution/src/lib.rs [hunk 4]
```diff
@@ -75,6 +75,8 @@ pub enum ExecutionError {
     #[cfg(any(feature = "wasmer", feature = "wasmtime"))]
     #[error(transparent)]
     WasmError(#[from] WasmExecutionError),
+    #[error(transparent)]
+    JoinError(#[from] tokio::task::JoinError),
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/lib.rs:79]: In principle, `tokio::task::JoinError` could fail for additional reasons in the future.

**Pull request merged**